### PR TITLE
add odos api

### PR DESCRIPTION
**Pull request by JohnnyGoodpick**

### CODE REVIEW

File: src/components/getOdosQuote.ts [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+	if (!fromAddress || fromAddress === '' || fromAddress === '0x') {
+		return new Err({
+			statusCode: 400,
+			data: 'Invalid from address',
+		});
+	}
+	const odosQuoteRequestParameters: OdosQuoteRequestParameters =
+		createQueryStringRequestObject(request);
+	const odosQuoteEndpoint = 'https://api.odos.xyz/sor/quote/v2';
+	const assembleEndpoint = 'https://api.odos.xyz/sor/assemble';
```
[the remark this answers is not in the document]
Comment (JohnnyGoodpick) [reply src/components/getOdosQuote.ts:98]: The sor/quote/v2 is to get a quote and a pathId
The /sor/assemble endpoint takes the pathId as argument and build the actual/real transaction and return the data to use in the tx

You cannot use the assemble endpoint without getting a pathId from the quote endpoint first.

image

https://docs.odos.xyz/api/quickstart/sor/

File: src/components/getOdosQuote.ts [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+	try {
+		const instance = axios.create();
+		instance.defaults.timeout = 5000;
```
[the remark this answers is not in the document]
Comment (JohnnyGoodpick) [reply src/components/getOdosQuote.ts:101]: I agree its too short for odos btw the same timeout value is used in the others getXXXQuote.ts file

File: src/components/getOdosQuote.ts [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+
+		if (r.status !== 200) {
+			throw r;
+		}
+
+		const quoteResponse = r.data as OdosQuoteResponse;
+		logger.debug('odos quote data', quoteResponse);
+		if (
+			!quoteResponse.pathId ||
+			quoteResponse.inAmounts[0] === '0' ||
```
[the remark this answers is not in the document]
Comment (JohnnyGoodpick) [reply src/components/getOdosQuote.ts:120]: I've seen cases where odos return '0' as quote on very low amount

File: src/components/getOdosQuote.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (0xelMag0) [reply src/components/getOdosQuote.ts:115]: I've read the difference after writing the comment, thank you for the explanation regardless.

File: src/components/getOdosQuote.ts [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+			quoteResponse.outAmounts[0] === '0'
+		) {
+			return new Err({
+				statusCode: 500,
+				data: 'Odos Quote failed',
+			});
+		}
+		// Fetch the built calldata for the quote
+		const assembleRequestParameters: OdosAssembleRequestParameters = {
```
[the remark this answers is not in the document]
Comment (0xelMag0) [reply src/components/getOdosQuote.ts:160]: Yes, we use the calldata, but only upon swapping & not when we fetch the quote.

[the remark this answers is not in the document]
Comment (JohnnyGoodpick) [reply src/components/getOdosQuote.ts:160]: The code has been modified to stop just before fetching the calldata when skipValidation is true, it is the case for ex in the swagger docs

**Pull request merged**